### PR TITLE
PQC: add CryptoCb support for PQC algorithms

### PR DESCRIPTION
**Pull request by Frauschi**
Hi there,

this PR introduces initial Crypto callback support for Post-Quantum algorithms (Kyber, Dilithium and Falcon). 

In our research project, we are creating a Post-Quantum enabled Secure Element, for which we need proper support from within the TLS library. 

I'm sure the current state of the integration misses some key areas for complete support (hence the draft state of the PR). However, I'm looking forward to your feedback regarding the overall direction of the implementation.

### CODE REVIEW

**Comment by dgarske** [comment]
Okay to test. Contributor agreement on file

**Comment by dgarske** [comment]
```[check-source-text] [2 of 7] [5c5f684888]
8 bit, trailing spaces or tabs:
./wolfcrypt/src/ext_kyber.c:435:    if ((ret == 0)
./wolfcrypt/src/falcon.c:86:    size_t localOutLen = 0;
weird control chars, hard tabs, CRs, trailing whitespace:
./wolfcrypt/src/ext_kyber.c:435:    if ((ret == 0)
overlong lines added:
/src/internal.c:27458         dilithiumKey = (dilithium_key*)XMALLOC(sizeof(dilithium_key), heap, DYNAMIC_TYPE_ECC);
/src/internal.c:27464             ret = wc_dilithium_init_label(dilithiumKey, (char*)data, heap, devId);
/src/internal.c:27481         falconKey = (falcon_key*)XMALLOC(sizeof(falcon_key), heap, DYNAMIC_TYPE_ECC);
/src/internal.c:27655                 *length = (word16)wc_dilithium_sig_size((dilithium_key*)ssl->hsKey);
/wolfcrypt/src/cryptocb.c:788 #if defined(HAVE_PQC) && (defined(WOLFSSL_WC_KYBER) || defined(HAVE_LIBOQS) || defined(HAVE_PQM4))
/wolfcrypt/src/cryptocb.c:814 int wc_CryptoCb_KyberEncapsulate(byte* ciphertext, word32 ciphertextLen, byte* sharedSecret,
/wolfcrypt/src/dilithium.c:899     byte privKey[DILITHIUM_MAX_PRV_KEY_SIZE], pubKey[DILITHIUM_MAX_PUB_KEY_SIZE];
/wolfcrypt/src/ext_kyber.c:440         ret = wc_CryptoCb_KyberEncapsulate(ct, ctlen, ss, KYBER_SS_SZ, rng, key);
/wolfssl/wolfcrypt/cryptocb.h:219         #if defined(HAVE_PQC) && (defined(WOLFSSL_WC_KYBER) || defined(HAVE_LIBOQS) || defined(HAVE_PQM4))
/wolfssl/wolfcrypt/cryptocb.h:548 #if defined(HAVE_PQC) && (defined(WOLFSSL_WC_KYBER) || defined(HAVE_LIBOQS) || defined(HAVE_PQM4))
/wolfssl/wolfcrypt/cryptocb.h:552 WOLFSSL_LOCAL int wc_CryptoCb_KyberEncapsulate(byte* ciphertext, word32 ciphertextLen,
/wolfssl/wolfcrypt/cryptocb.h:555 WOLFSSL_LOCAL int wc_CryptoCb_KyberDecapsulate(const byte* ciphertext, word32 ciphertextLen,
/wolfssl/wolfcrypt/cryptocb.h:563 WOLFSSL_LOCAL int wc_CryptoCb_DilithiumSign(const byte* in, word32 inlen, byte* out,
/wolfssl/wolfcrypt/cryptocb.h:569 WOLFSSL_LOCAL int wc_CryptoCb_DilithiumCheckPrivKey(dilithium_key* key, const byte* pubKey,
/wolfssl/wolfcrypt/cryptocb.h:577 WOLFSSL_LOCAL int wc_CryptoCb_FalconSign(const byte* in, word32 inlen, byte* out,
/wolfssl/wolfcrypt/cryptocb.h:583 WOLFSSL_LOCAL int wc_CryptoCb_FalconCheckPrivKey(falcon_key* key, const byte* pubKey,
/wolfssl/wolfcrypt/types.h:1173     #if defined(HAVE_PQC) && (defined(WOLFSSL_WC_KYBER) || defined(HAVE_LIBOQS) || defined(HAVE_PQM4))
    check-source-text fail_AB```

**Comment by Frauschi** [comment]
@dgarske Formatting issues should be addressed now

File: src/internal.c [hunk 1]
```diff
@@ -27451,6 +27451,55 @@ int CreateDevPrivateKey(void** pkey, byte* data, word32 length, int hsType,
         }
 #endif
     }
+    else if (hsType == DYNAMIC_TYPE_DILITHIUM) {
+#if defined(HAVE_PQC) && defined(HAVE_DILITHIUM)
+        dilithium_key* dilithiumKey;
+
+        dilithiumKey = (dilithium_key*)XMALLOC(sizeof(dilithium_key), heap,
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply src/internal.c:27458]: fixed

File: src/internal.c [hunk 2]
```diff
@@ -27451,6 +27451,55 @@ int CreateDevPrivateKey(void** pkey, byte* data, word32 length, int hsType,
...
+        else {
+            XFREE(dilithiumKey, heap, DYNAMIC_TYPE_ECC);
+        }
+#endif
+    }
+    else if (hsType == DYNAMIC_TYPE_FALCON) {
+#if defined(HAVE_PQC) && defined(HAVE_FALCON)
+        falcon_key* falconKey;
+
+        falconKey = (falcon_key*)XMALLOC(sizeof(falcon_key), heap,
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply src/internal.c:27483]: fixed

File: wolfcrypt/src/dilithium.c [hunk 3]
```diff
@@ -200,15 +236,103 @@ int wc_dilithium_verify_msg(const byte* sig, word32 sigLen, const byte* msg,
...
+    if (ret == 0 && (len < 0 || len > DILITHIUM_MAX_ID_LEN))
+        ret = BUFFER_E;
+
+    if (ret == 0)
+        ret = wc_dilithium_init_ex(key, heap, devId);
+    if (ret == 0 && id != NULL && len != 0) {
+        XMEMCPY(key->id, id, (size_t)len);
+        key->idLen = len;
+    #ifdef WOLFSSL_SE050
+        /* Set SE050 ID from word32, populate ecc_key with public from SE050 */
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/dilithium.c:296]: TBH I just copied the `_id` and `_label` code from the existing RSA and ECC stuff, where this SE050 is all around. Therefore, I placed it inside the PQC stuff, too. If it isn't necessary here, I gladly remove it. It is definitely not tested in any way on my side.

File: wolfcrypt/src/cryptocb.c [hunk 4]
```diff
@@ -785,6 +785,314 @@ int wc_CryptoCb_Ed25519Verify(const byte* sig, word32 sigLen,
 }
 #endif /* HAVE_ED25519 */
 
+#if defined(HAVE_PQC) && defined(WOLFSSL_HAVE_KYBER)
+int wc_CryptoCb_MakeKyberKey(WC_RNG* rng, int keySize, KyberKey* key)
```
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/cryptocb.c]: I updated the commit to reflect your proposal. We now have a set of CryptoCb functions for PQC KEMs and one for PQC Signature algorithms.

[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/cryptocb.c]: Ok, another update, as I wasn't happy with the previous fix. I now defined two new “types” `WC_PQC_KEM_TYPE_x` and `WC_PQC_SIG_TYPE_x` and one common set of CryptoCb methods for each. This should be a future-proof design for easy integration of new algorithms. Looking forward to your feedback :)

**Comment by dgarske** [comment]
Retest this please

**Review by dgarske — DISMISSED** [review]
Very nice work! @anhu will you review as well?

**Review by anhu — APPROVED** [review]

**Review by douzzer — CHANGES_REQUESTED** [review]
a quick round of testing turned up some issues:
```
[pq-all] [1 of 4] [ac7135e869]
    autogen.sh ac7135e869...   real 0m15.788s  user 0m14.032s  sys 0m0.863s
    configure...   real 0m14.393s  user 0m6.724s  sys 0m9.220s
    build...ac7135e869 (<tobias.frauenschlaeger@oth-regensburg.de> 2023-12-20 16:22:07 +0100 789) int wc_CryptoCb_PqcKemGetDevId(int type, void* key)
wolfcrypt/src/cryptocb.c:789:5: error: no previous prototype for ‘wc_CryptoCb_PqcKemGetDevId’ [-Werror=missing-prototypes]
  789 | int wc_CryptoCb_PqcKemGetDevId(int type, void* key)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
36d7bcc5b8 (<tobias.frauenschlaeger@oth-regensburg.de> 2023-12-06 17:44:20 +0100 909) int wc_CryptoCb_PqcSigGetDevId(int type, void* key)
wolfcrypt/src/cryptocb.c:909:5: error: no previous prototype for ‘wc_CryptoCb_PqcSigGetDevId’ [-Werror=missing-prototypes]
  909 | int wc_CryptoCb_PqcSigGetDevId(int type, void* key)
      |     ^~~~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
make[2]: *** [Makefile:7331: wolfcrypt/src/src_libwolfssl_la-cryptocb.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:8018: all-recursive] Error 1
make: *** [Makefile:4747: all] Error 2
   real 0m9.747s  user 0m26.065s  sys 0m1.773s
    pq-all fail_build
    failed config: '--enable-all' '--enable-testcert' '--enable-srtp' '--enable-intelasm' '--enable-sp-asm' '--with-liboqs' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK -pedantic'
[pq-all-clang-tidy] [2 of 4] [ac7135e869]
    configure...   real 0m20.143s  user 0m10.900s  sys 0m10.577s
    build...ac7135e869 (<tobias.frauenschlaeger@oth-regensburg.de> 2023-12-20 16:22:07 +0100 107)     key->devCtx = NULL;
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:107:17: warning: Access to field 'devCtx' results in a dereference of a null pointer (loaded from variable 'key') [clang-analyzer-core.NullDereference]
107 |     key->devCtx = NULL;
|     ~~~         ^
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 80)     if (key == NULL) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:80:9: note: Assuming 'key' is equal to NULL
80 |     if (key == NULL) {
|         ^~~~~~~~~~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 80)     if (key == NULL) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:80:5: note: Taking true branch
80 |     if (key == NULL) {
|     ^
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 83)     if (ret == 0) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:83:9: note: 'ret' is not equal to 0
83 |     if (ret == 0) {
|         ^~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 83)     if (ret == 0) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:83:5: note: Taking false branch
83 |     if (ret == 0) {
|     ^
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 98)     if (ret == 0) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:98:9: note: 'ret' is not equal to 0
98 |     if (ret == 0) {
|         ^~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 98)     if (ret == 0) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:98:5: note: Taking false branch
98 |     if (ret == 0) {
|     ^
ac7135e869 (<tobias.frauenschlaeger@oth-regensburg.de> 2023-12-20 16:22:07 +0100 107)     key->devCtx = NULL;
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:107:17: note: Access to field 'devCtx' results in a dereference of a null pointer (loaded from variable 'key')
107 |     key->devCtx = NULL;
|     ~~~         ^
ac7135e869 (<tobias.frauenschlaeger@oth-regensburg.de> 2023-12-20 16:22:07 +0100 546)     if (key->devId != INVALID_DEVID)
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:546:9: warning: Access to field 'devId' results in a dereference of a null pointer (loaded from variable 'key') [clang-analyzer-core.NullDereference]
546 |     if (key->devId != INVALID_DEVID)
|         ^~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 534)     if ((key == NULL) || (ss == NULL) || (ct == NULL)) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:534:10: note: Assuming 'key' is equal to NULL
534 |     if ((key == NULL) || (ss == NULL) || (ct == NULL)) {
|          ^~~~~~~~~~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 534)     if ((key == NULL) || (ss == NULL) || (ct == NULL)) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:534:23: note: Left side of '||' is true
534 |     if ((key == NULL) || (ss == NULL) || (ct == NULL)) {
|                       ^
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 537)     if (ret == 0) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:537:9: note: 'ret' is not equal to 0
537 |     if (ret == 0) {
|         ^~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 537)     if (ret == 0) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:537:5: note: Taking false branch
537 |     if (ret == 0) {
|     ^
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 540)     if ((ret == 0) && (len != ctlen)) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:540:10: note: 'ret' is not equal to 0
540 |     if ((ret == 0) && (len != ctlen)) {
|          ^~~
a2635be9e6 (<anthony@wolfssl.com> 2022-09-09 11:41:30 -0400 540)     if ((ret == 0) && (len != ctlen)) {
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:540:20: note: Left side of '&&' is false
540 |     if ((ret == 0) && (len != ctlen)) {
|                    ^
ac7135e869 (<tobias.frauenschlaeger@oth-regensburg.de> 2023-12-20 16:22:07 +0100 546)     if (key->devId != INVALID_DEVID)
/tmp/tmp.4346_10542/wolfssl_test_workdir.19929/wolfssl/wolfcrypt/src/ext_kyber.c:546:9: note: Access to field 'devId' results in a dereference of a null pointer (loaded from variable 'key')
546 |     if (key->devId != INVALID_DEVID)
|         ^~~
Suppressed 38 warnings (38 in non-user code).
make[2]: *** [Makefile:7170: wolfcrypt/src/src_libwolfssl_la-ext_kyber.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
make[1]: *** [Makefile:8018: all-recursive] Error 1
make: *** [Makefile:4747: all] Error 2
   real 1m36.743s  user 9m26.285s  sys 0m12.829s
    pq-all-clang-tidy fail_analytic_build
```

**Comment by Frauschi** [comment]
> a quick round of testing turned up some issues:
> 
Should all be addressed now, thanks for the feedback. 

Out of interest, is it possible to run these tests locally on my side, too?

File: wolfcrypt/src/dilithium.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Frauschi) [reply wolfcrypt/src/dilithium.c]: I just looked again at the SE050 stuff. As this secure element doesn't support any Post Quantum algorithms, I removed these special cases in the code completely.

**Comment by dgarske** [comment]
okay to test / retest this please

**Comment by douzzer** [comment]
@Frauschi the first of the two test scenarios in my comment, `pq-all`, is just a regular configure-build-make-check.  The test harness prints the `./configure` args to reproduce the scenario.  Here,
```
failed config: '--enable-all' '--enable-testcert' '--enable-srtp' '--enable-intelasm' '--enable-sp-asm' '--with-liboqs' 'CPPFLAGS=-DNO_WOLFSSL_CIPHER_SUITE_TEST -DWOLFSSL_OLD_PRIME_CHECK -pedantic'
```
So that's how you'd test that locally.  Easy.

The 2nd scenario, `pq-all-clang-tidy`, is a different story.  In order to build under `clang-tidy` we have to do all sorts of tricky scripting and meta-configuration, which as @dgarske says is part of our internal testing system.

Building and checking with sanitizers may find some of the defects reported by `clang-tidy` and other static analyzers, and you can do that by just adding to the `./configure` args something like
```
CFLAGS='-g -fno-omit-frame-pointer -fsanitize=address,pointer-subtract,leak,undefined,float-cast-overflow,float-divide-by-zero,bounds-strict -fsanitize-recover=all' LDFLAGS='-g -fno-omit-frame-pointer -fsanitize-recover=all -fsanitize=address,pointer-subtract,leak,undefined,float-cast-overflow,float-divide-by-zero,bounds-strict'
```

**Review by douzzer — DISMISSED** [review]
Looks great.  Passed `wolfssl-multi-test.sh ... pq-all pq-all-clang-tidy pq-all-sanitizer pq-all-cppcheck`

File: wolfcrypt/src/cryptocb.c [hunk 5]
```diff
@@ -785,6 +785,284 @@ int wc_CryptoCb_Ed25519Verify(const byte* sig, word32 sigLen,
...
+int wc_CryptoCb_MakePqcKemKey(WC_RNG* rng, int type, int keySize, void* key)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+    CryptoCb* dev;
+
+    if (key == NULL)
+        return ret;
+
+    /* get devId */
+    int devId = wc_CryptoCb_PqcKemGetDevId(type, key);
```
Comment (dgarske) [inline wolfcrypt/src/cryptocb.c:815]: Mixed declaration. Variables must be declared at top of function or brace {} section.

File: wolfcrypt/src/cryptocb.c [hunk 6]
```diff
@@ -785,6 +785,284 @@ int wc_CryptoCb_Ed25519Verify(const byte* sig, word32 sigLen,
...
+    void* key)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+    CryptoCb* dev;
+
+    if (key == NULL)
+        return ret;
+
+    /* get devId */
+    int devId = wc_CryptoCb_PqcKemGetDevId(type, key);
```
Comment (dgarske) [inline wolfcrypt/src/cryptocb.c:848]: Same. Several more below

File: wolfcrypt/src/cryptocb.c [hunk 7]
```diff
@@ -785,6 +785,284 @@ int wc_CryptoCb_Ed25519Verify(const byte* sig, word32 sigLen,
...
+int wc_CryptoCb_PqcDecapsulate(const byte* ciphertext, word32 ciphertextLen,
+    byte* sharedSecret, word32 sharedSecretLen, int type, void* key)
+{
+    int ret = CRYPTOCB_UNAVAILABLE;
+    CryptoCb* dev;
+
+    if (key == NULL)
+        return ret;
+
+        /* get devId */
```
Comment (dgarske) [inline wolfcrypt/src/cryptocb.c:882]: Indent off. 4 spaces please

**Review by dgarske — CHANGES_REQUESTED** [review]
Also please add `liboqs.h` to the make install. It needs added to the include.am with EXTRA_DIST I believe.

```
./configure --with-liboqs; make all; sudo make  install ; find /usr/local/include | grep liboqs.h```

**Comment by Frauschi** [comment]
@douzzer thanks very much for the thorough response, will definitely get my hands on the sanitizer stuff 😄

**Comment by douzzer** [comment]
Re @dgarske 's comment on missing installation of `liboqs.h`, I think what you need to do is, in `wolfssl/src/wolfssl/wolfssl/wolfcrypt/include.am`, move `wolfssl/wolfcrypt/port/liboqs/liboqs.h` from `noinst_HEADERS` to `nobase_include_HEADERS`, and make it conditional on `BUILD_LIBOQS` as it is for `BUILD_CRYPTOAUTHLIB` and the various other optional/target-specific headers there.

I think your inclusion of it in `noinst_HEADERS` was functionally equivalent to `EXTRA_DIST`, i.e. the new header was getting properly included in the dist archive (we test for that), so you just need to adjust so that it gets installed by `make install`.

File: wolfcrypt/src/cryptocb.c (hunk 5 above)
Comment (Frauschi) [reply wolfcrypt/src/cryptocb.c]: fixed

File: wolfcrypt/src/cryptocb.c (hunk 6 above)
Comment (Frauschi) [reply wolfcrypt/src/cryptocb.c]: fixed all (hopefully)

File: wolfcrypt/src/cryptocb.c (hunk 7 above)
Comment (Frauschi) [reply wolfcrypt/src/cryptocb.c]: fixed

**Comment by Frauschi** [comment]
> Also please add `liboqs.h` to the make install. It needs added to the include.am with EXTRA_DIST I believe.

Should also be fixed now. Thanks @douzzer for the hints about how to fix it :)

**Comment by dgarske** [comment]
okay to test

**Review by dgarske — APPROVED** [review]

**Pull request merged**